### ISSUE TITLE
GPU kernel for forall loops might hang in the presense of syncThreads primitive.

### ISSUE THREAD

**Issue by DanilaFe** [chapel-lang/chapel#21460]
Currently, GPUized forall loops have kernels that include implicit "bounds checking" using the thread and block ID on the device. For instance, for the following `forall` loop:
```Chapel
forall x in 0..N {
  f(x);
}
```
The generated kernels looks _kind of_ like this:
```Chapel
proc chpl_kernel() {
   var idx = blockId * blockSize + threadId;
   if idx > N {
     return;
   }
   f(x);
}
```
This is an established pattern on the GPU; the kernel is executed on many threads, and some of these threads may be outside of the range of the data we want to be operating on. All is well.

The (possible) trouble is in the use of the `syncThreads` primitive. The primitive acts like a barrier; conceptually, all threads must hit the barrier before continuing. In practice, on NVIDIA (prior to compute capability 7.0), barrier completion was tracked on a per-warp basis (any one thread in a warp could hit the barrier and thereby increment its count by 32). Either way, it is sufficient for one thread / one warp to not call `syncThreads` to create a deadlock. [^1] One indication of this behavior is the following skeleton of the matrix transpose kernel from the [NVIDIA linear algebra examples](https://www.nvidia.com/content/cudazone/cuda_sdk/Linear_Algebra.html#transpose):

```CUDA
	if((xIndex < width) && (yIndex < height)) { /* ... */ }
	__syncthreads();
	if((xIndex < height) && (yIndex < width)) { /* ... */ }
```

Notice that `__syncThreads()` is invoked unconditionally.

It thus seems to me like with the current implementation of the for-loop kernels, it would be possible to write some fairly trivial foreach loops that don't halt (as long as the _non-exited_ condition is not present, as would be the case prior to Volta for CUDA, and possibly for other GPU vendors). One example of such a loop would be:

```chpl
forall 1..(chplBlockSize / 2) do __primitive("gpu syncThreads");
```

Note that I couldn't get any version of this loop to hang on either a Tesla P100 or a T4, but I think the problem is there in principle.

Things I have tried to make the GPU halt to no avail:
```Chapel
use GPU;
use GPUDiagnostics;

on here.gpus[0] {
  startVerboseGPU();
  var A: [0..128] int;
  writeln(A.size);
  // If warps are in multiples of 32, then A is covered by 4 warps,
  // while the remaining 12 warps in the default block size (512) should all exit early without calling syncThreads.
  // Even if barriers are tracked at a warp granularity, this could hang.
  foreach a in A {
    assertOnGpu();
    a += 1;
    __primitive("gpu syncThreads");
    a += 1;
  }
  writeln(A);
  // If barriers are tracked at a per-thread (including exited threads) granularity, this should be sufficient,
  // as it will leave one thread in the block to do nothing (and thus not call syncThreads).
  foreach i in 0..<(512 - 1) do __primitive("gpu syncThreads");
  // This is more of a mean-spirited example; all threads should execute, but only half of them should
  // sync.
  foreach i in 0..<512 /* assume blockSize of 512 */ {
    if i < 256 then __primitive("gpu syncThreads");
  }
  stopVerboseGPU();
}
```

Likely the saving grace here is that exited threads may count as barrier completions, thus the early-exiting threads don't need to participate in `syncThreads`. 

[^1]: As of CC 7.x, apparently the barrier only requires all _non-exited_ threads to complete; see https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#synchronization-functions and https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#independent-thread-scheduling.

**Comment by bradcray** [comment]
I'm definitely not the GPU expert here, so it may be that targeting GPUs changes something important, but a superficial reaction to the loop:

```chapel
forall 1..(chplBlockSize / 2) do __primitive("gpu syncThreads");
```

is that putting something that's barrier-like into a `forall` or `foreach` seems like it contradicts the required/desired semantics of "can be run by any number of threads in an order-independent manner."  Normally, if I wanted to write the pattern:

```chapel
forall i in 1..n [
  a;
  barrier();
  b;
}
```

in Chapel, I'd split it into two forall loops (since the termination of the first would imply the barrier) or convert it into a coforall (which doesn't have the "order independent" implication, so supports barriers legally).  Making it a coforall doesn't seem like a good option for a GPU computation, though.

I'll also mention that we've had users request for the forall-with-barrier pattern above to work, where their hope was that each task, upon hitting the barrier, would continue and do its other iterations' a work, then enter the barrier, then do b for all of its iterations.  That's never been implemented, though.

**Comment by DanilaFe** [comment]
Yes, two for-loops would work -- unless you're writing to block shared memory. This is why thread synchronization is needed; in the transpose benchmark, a thread writes to `sharedMem[y][x]`, but reads from `sharedMem[x][y]`. If the other thread, the one that that _writes_ to `sharedMem[x][y]`, doesn't complete in time for the first thread's read, you'll get weird results. But the shared memory only exists within the kernel; thus, splitting into two for-loops would not work.

**Comment by stonea** [comment]
> As of CC 7.x, apparently the barrier only requires all non-exited threads to complete; see docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#synchronization-functions and docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#independent-thread-scheduling.

So this would imply that this just won't be an issue for NVIDIA going forward, right?

If so, I'm not sure to what extent we care about supporting older GPUs.  If there's some way we could trivially fix this (without impacting performance) I'd say we should go for that.

That said I can't think of a trivial fix, and since we've been unable to reproduce the issue on our end, I think it makes sense to:

- Add something to the technote that for NVIDIA, we require compute capability >= 7.x.
  - And then do some sort of check for this at runtime in our CUDA-based runtime lib
    (we could create a separate GH issue to add this later).
- Add a test that we would expect to timeout/fail if this is an issue.
  - And then in that tests add some conspicuous comments explaining what issue
    the test is meant to catch and link back to this GH issue.
- And then close this issue.

**Comment by e-kayrakli** [comment]
Going through some of the documentation, it looks like what we're doing is fine in implementation but not really in documentation. IOW, we are probably slipping through cracks right now on P100s.

@stonea Note that we default to compute capability 6.0 and P100s are 6.0. So, it may be early to call 6.0 unsupported today. Though I don't know what we should do at the moment. I'd support

> Add a test that we would expect to timeout/fail if this is an issue.

with a lot of trials to see if it ever deadlocks in nightlies.

---

@bradcray as you reacted to Daniel's response with a confused emoji, I'll explain the difference in my own words:

The barrier here is not a global one, it'll only barrier a subset of parallel units. As such it doesn't correspond that closely with our barriers today. As Daniel mentioned, your 2-loop strategy is not really applicable here as that would imply everyone doing `a()` first. And only after that point they do `b()`. But instead we want to wait for all parallel units within the same group/team/warp to finish `a()` before the same set of units moving forward with `b()`.

To add to the complexity here, shared memory allocations (typically go together with `__syncthreads()`) can be thought of as stack allocations. Considering the 2-loop scenario, they will be deallocated when the first loop ends.

**Comment by bradcray** [comment]
> The barrier here is not a global one, it'll only barrier a subset of parallel units. As such it doesn't correspond that closely with our barriers today.

allLocalesBarrier() is "global" by some definition, but our normal barrier isn't, is it?  (e.g., I could create a barrier that sync'd between two tasks on a single locale without involving any other tasks or locales).

> As Daniel mentioned, your 2-loop strategy is not really applicable here as that would imply everyone doing a() first. And only after that point they do b(). But instead we want to wait for all parallel units within the same group/team/warp to finish a() before the same set of units moving forward with b().

I'm still not understanding the distinction.  Is it simply "everyone" (by some definition of everyone that's not clear to me) vs. "some things"?

**Comment by e-kayrakli** [comment]
> allLocalesBarrier() is "global" by some definition, but our normal barrier isn't, is it? (e.g., I could create a barrier that sync'd between two tasks on a single locale without involving any other tasks or locales).

Correct, but this one is more implicit: you can't choose who is participating in the barrier.

> I'm still not understanding the distinction. Is it simply "everyone" (by some definition of everyone that's not clear to me) vs. "some things"?

A non-gpu sketch of the problem that uses some currently-imaginary features:

```chapel
forall i in 1..n {
  var sharedArr = createSharedArray(...); // you can imagine this as a per-locale local array
                                          // that is shared across tasks locally
  
  if taskId == 0 {  // taskId is the local ID
    // populate sharedArr
  }

  barrierInQuestion(); // make sure taskId!=0 waits for taskId==0

  work(sharedArr);
}
```

You don't really want `barrierInQuestion()` to cause Locale 1's tasks to wait for Locale 0's. Nor you could break the loop in two because then `sharedArr` will not be in scope of the second loop.

These are real issues in the GPU-land today, but they aren't outside of it. With new forall features we have been discussing to enable SPMD-like programming, what I sketched above will be more real globally. Which.. re-reading your initial comment, you've captured in a different way:

> in Chapel, I'd split it into two forall loops (since the termination of the first would imply the barrier) or convert it into a coforall (which doesn't have the "order independent" implication, so supports barriers legally). Making it a coforall doesn't seem like a good option for a GPU computation, though.

IOW: `coforall` is not an option for GPUs, and for non-GPU cases, we want to raise the threshold of falling back to `coforall`s.

**Comment by bradcray** [comment]
Restating the original concern another way:

Do you have a better sense than I do as to what makes it legal/reasonable to put a syncThreads into a foreach / forall loop without violating the "order-independent"- / "can be run by any number of threads"-style properties that we've historically associated with them, and how we'd explain that to a user?  For example, do our traditional descriptions of the loop forms and what they imply need to evolve and be updated in the documentation (and my head)?  Or is there something specific about syncThreads that makes it not violate the properties?  That's what remains unclear to me.

(This question is also related to https://github.com/chapel-lang/chapel/issues/19153, where I worry that we haven't made it clear what is / is not permitted within foreach loops, and currently permit some expressions that are not legal within foralls—perhaps incorrectly/unintentionally).

**Comment by e-kayrakli** [comment]
> Do you have a better sense than I do as to what makes it legal/reasonable to put a syncThreads into a foreach / forall loop without violating the "order-independent"- / "can be run by any number of threads"-style properties that we've historically associated with them, and how we'd explain that to a user? For example, do our traditional descriptions of the loop forms and what they imply need to evolve and be updated in the documentation (and my head)? Or is there something specific about syncThreads that makes it not violate the properties? That's what remains unclear to me.

I see. Without accounting for how we describe these concerns in documentation etc. I am not worried about them because my take is:

> "order-independent"

means iterations can start and finish in any order without changing the state of the program at the end of the loop. Another way to express how I think about this is, the induction variable's position within the iteration space is at any given point in time doesn't tell you anything about how much of the iteration is done or whether a specific iteration is completed. I don't see synchronization breaking that property.

In the same vein, if we could some sort of global barriers work in non-gpu foralls, I wouldn't see them as breaking forall's properties.

>  "can be run by any number of threads"

I am not sure how to interpret this property. But probably what we'd need to do here is to distinguish between thread-level parallelism and vector-lane-level parallelism (almost wanted to say instruction-level, but that's something else..). But the terminology is a bit confusing here as we see GPU threads to be equivalent to vector lanes (and were discussing the term "fiber" to describe this). So, "syncThreads" here is actually a barrier across vector lanes.
